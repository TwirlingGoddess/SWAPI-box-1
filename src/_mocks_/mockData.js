const mockVehicleApiResponse = {
  count: 39,
  next: 'https://swapi.co/api/vehicles/?page=2',
  previous: null,
  results: [
    {
      cargo_capacity: '50000',
      consumables: '2 months',
      cost_in_credits: '150000',
      created: '2014-12-10T15:36:25.724000Z',
      crew: '46',
      edited: '2014-12-22T18:21:15.523587Z',
      films: ['https://swapi.co/api/films/5/', 
        'https://swapi.co/api/films/1/'],
      length: '36.8',
      manufacturer: 'Corellia Mining Corporation',
      max_atmosphering_speed: '30',
      model: 'Digger Crawler',
      name: 'Sand Crawler',
      passengers: '30',
      pilots: [],
      url: 'https://swapi.co/api/vehicles/4/',
      vehicle_class: 'wheeled'
    }
  ]
};

const mockPeopleApiResponse = {
  count: 87,
  next: 'https://swapi.co/api/people/?page=2',
  previous: null,
  results: [
    {
      birth_year: '19BBY',
      created: '2014-12-09T13:50:51.644000Z',
      edited: '2014-12-20T21:17:56.891000Z',
      eye_color: 'blue',
      films: ['https://swapi.co/api/films/2/',
        'https://swapi.co/api/films/6/',
        'https://swapi.co/api/films/3/'],
      gender: 'male',
      hair_color: 'blond',
      height: '172',
      homeworld: 'https://swapi.co/api/planets/1/',
      mass: '77',
      name: 'Luke Skywalker',
      skin_color: 'fair',
      species: ['https://swapi.co/api/species/1/'],
      starships: ['https://swapi.co/api/starships/12/',
        'https://swapi.co/api/starships/22/'],
      url: 'https://swapi.co/api/people/1/',
      vehicles: ['https://swapi.co/api/vehicles/14/',
        'https://swapi.co/api/vehicles/30/']
    }
  ]
};

const mockParseData = [
  {
    birth_year: '19BBY',
    created: '2014-12-09T13:50:51.644000Z',
    edited: '2014-12-20T21:17:56.891000Z',
    eye_color: 'blue',
    films: ['https://swapi.co/api/films/2/',
      'https://swapi.co/api/films/6/',
      'https://swapi.co/api/films/3/'],
    gender: 'male',
    hair_color: 'blond',
    height: '172',
    homeworld: 'https://swapi.co/api/planets/1/',
    mass: '77',
    name: 'Luke Skywalker',
    skin_color: 'fair',
    species: ['https://swapi.co/api/species/1/'],
    starships: ['https://swapi.co/api/starships/12/', 
      'https://swapi.co/api/starships/22/'],
    url: 'https://swapi.co/api/people/1/',
    vehicles: ['https://swapi.co/api/vehicles/14/', 
      'https://swapi.co/api/vehicles/30/']
  }
];

const mockPlanetApiResponse = {
  count: 61,
  next: 'https://swapi.co/api/planets/?page=2',
  previous: null,
  results: [
    {
      climate: 'temperate',
      created: '2014-12-10T11:35:48.479000Z',
      diameter: '12500',
      edited: '2014-12-20T20:58:18.420000Z',
      films: ['https://swapi.co/api/films/6/', 
        'https://swapi.co/api/films/1/'],
      gravity: '1 standard',
      name: 'Alderaan',
      orbital_period: '364',
      population: '2000000000',
      residents: ['https://swapi.co/api/people/5/',
        'https://swapi.co/api/people/68/',
        'https://swapi.co/api/people/81/'],
      rotation_period: '24',
      surface_water: '40',
      terrain: 'grasslands, mountains',
      url: 'https://swapi.co/api/planets/2/'
    }
  ]
};

const mockPeopleObject = [{'Homeworld': undefined, 
  'Name': 'Luke Skywalker', 
  'Population': undefined, 
  'Specie': undefined, 'id': 
'people0', 
  'keyList': 'people'
}];

const mockPeopleObject2 = {
  'Homeworld': undefined, 
  'Name': 'Leia', 
  'Population': undefined, 
  'Specie': undefined, 'id': 
'people2', 
  'keyList': 'people'
};

const mockFavorites = [
  {'Homeworld': undefined, 
    'Name': 'Luke Skywalker', 
    'Population': undefined, 
    'Specie': undefined, 'id': 
'people0', 
    'keyList': 'people'
  },
  {
    'Homeworld': undefined, 
    'Name': 'Leia', 
    'Population': undefined, 
    'Specie': undefined, 'id': 
  'people2', 
    'keyList': 'people'
  }
];

const filmCrawlResponse = [{
  crawl: 'Film Crawl',
  title: 'Movie Title',
  date: '1982-09-18',
  opening_crawl: 'There is unrest in the Galactic'},
{
  crawl: 'Film Crawl',
  title: 'Movie Title',
  date: '1982-09-18',
  opening_crawl: 'There is unrest in the Galactic'
}];


const filmCrawl = {
  crawl: 'Film Crawl',
  title: 'Movie Title',
  date: '1982-09-18',
  opening_crawl: 'There is unrest in the Galactic'
};

const vehicleData = [{
  Class: 'sail barge',
  id: 'vehicles0',
  Model: 'Modified Luxury Sail Barge',
  Name: 'Sail barge',
  Passengers: '500',
  keyList: 'vehicles'
}];

const planetData = [{
  Climate: 'temperate',
  id: 'planets0',
  Name: 'Alderaan',
  Residents: 'Alderaan',
  Population: '2000000000',
  Terrain: 'grasslands, mountains',
  keyList: 'planets'
}];

const parsedVehicleData = [
  {
    'name': 'Sail barge',
    'model': 'Modified Luxury Sail Barge',
    'manufacturer': 'Ubrikkian Industries Custom Vehicle Division',
    'cost_in_credits': '285000',
    'length': '30',
    'max_atmosphering_speed': '100',
    'crew': '26',
    'passengers': '500',
    'cargo_capacity': '2000000',
    'consumables': 'Live food tanks',
    'vehicle_class': 'sail barge',
    'pilots': [],
    'films': [
      'https://swapi.co/api/films/3/'
    ],
    'created': '2014-12-18T10:44:14.217000Z',
    'edited': '2014-12-22T18:21:15.807906Z',
    'url': 'https://swapi.co/api/vehicles/24/'
  }
];

const parsedPlanetData = [
  {
    climate: 'temperate',
    created: '2014-12-10T11:35:48.479000Z',
    diameter: '12500',
    edited: '2014-12-20T20:58:18.420000Z',
    films: ['https://swapi.co/api/films/6/', 
      'https://swapi.co/api/films/1/'],
    gravity: '1 standard',
    name: 'Alderaan',
    orbital_period: '364',
    population: '2000000000',
    residents: ['https://swapi.co/api/people/5/',
      'https://swapi.co/api/people/68/',
      'https://swapi.co/api/people/81/'],
    rotation_period: '24',
    surface_water: '40',
    terrain: 'grasslands, mountains',
    url: 'https://swapi.co/api/planets/2/'
  }
];

export { mockPlanetApiResponse, 
  mockVehicleApiResponse, 
  mockPeopleApiResponse, 
  mockPeopleObject, 
  mockPeopleObject2,
  mockFavorites,
  filmCrawl,
  filmCrawlResponse,
  mockParseData,
  vehicleData,
  planetData,
  parsedVehicleData,
  parsedPlanetData
};

