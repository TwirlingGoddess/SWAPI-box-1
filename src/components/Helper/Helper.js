import React from 'react';

export const cleanDataFunc = (dataObject) => {
  if (dataObject.next.includes('people')) {
    return peopleObject(dataObject.results);
  }
  if (dataObject.next.includes('planets')) {
    return planetObject(dataObject.results);
  }
  if (dataObject.next.includes('vehicles')) {
    return vehicleObject(dataObject.results);
  }
}

export const peopleObject = async (parsedData) => {
  const unresolvedPromises = parsedData.map( async(person,index) => {
    const keyList = 'people';
    const homeworld = await nestedFetch(person.homeworld);
    const homeworldName = homeworld.name;
    const homeworldPopulation = homeworld.population;
    const specieObj = await nestedFetch(person.species);
    const specie = specieObj.name;
    const name = person.name;
    const id = keyList + index;
    return {id, keyList, Homeworld: homeworldName, Population: homeworldPopulation, Specie: specie, Name:name};
  });
  return Promise.all(unresolvedPromises);
}


export const planetObject = async (parsedData) => {
  const unresolvedPromises = parsedData.map(async(planet,index) => {
    const keyList = 'planets';  
    const planetName = planet.name;
    const planetTerrain = planet.terrain;
    const planetPopulation = planet.population;
    const planetClimate = planet.climate;
    const residents = await residentsFetch(planet.residents);
    const id = keyList + index;
    return {id, keyList, Name: planetName, Terrain: planetTerrain, Population: planetPopulation, Climate: planetClimate, Residents: residents};
  });
  return Promise.all(unresolvedPromises);
}

export const vehicleObject = (parsedData) => {
  const vehicleArray = parsedData.map((vehicle, index) => {
    const keyList = 'vehicles'; 
    const vehicleName = vehicle.name;
    const vehicleModel = vehicle.model;
    const vehicleClass = vehicle.vehicle_class;
    const numberOfPassengers = vehicle.passengers;
    const id = keyList + index;
    return {id, keyList, Name: vehicleName, Model: vehicleModel, Class: vehicleClass, Passengers:numberOfPassengers};
  });
  return vehicleArray;
}

export const nestedFetch = async (url) => {
  const fetchURL= await fetch(url);
  const parseObject= await fetchURL.json();
  return parseObject;
}

export const residentsFetch = async (residentsUrls) => {
  const unresolvedPromises = residentsUrls.map(async(url) => {
    const residentURLfetch = await fetch(url);
    const parseResidents = await residentURLfetch.json();
    return parseResidents.name;
  });
  return Promise.all(unresolvedPromises);
}

export const makeApiCall = async (category) => {
  const url = `https://swapi.co/api/${category}`;
  return apiCall(url)
}

 export const apiCall = async (url) => {
  const fetchURL = await fetch(url);
  const parseObject = await fetchURL.json();
  const cleanData = await cleanDataFunc(parseObject);
  sendToLocalStorage(url, cleanData);
  return cleanData;
 }

export const sendToLocalStorage = (key, selectedData) => {
  localStorage.setItem(key, JSON.stringify(selectedData));
}

export const getFromLocalStorage = (key) => {
  return JSON.parse(localStorage.getItem(key));
}
