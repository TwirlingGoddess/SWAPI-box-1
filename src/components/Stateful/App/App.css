.App {
  text-align: center;
  background-color: black;
  height: 100vh;
  width: 100%;
  background-image: url("../../../assets/star_wars_kyloren_ren.gif");
  background-repeat: no-repeat;
  background-size: cover;
  font-family: 'Nanum Gothic Coding', monospace; }

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg); }
  to {
    transform: rotate(360deg); } }

body::-webkit-scrollbar {
  width: 1em; }

body::-webkit-scrollbar-track {
  -webkit-box-shadow: inset 0 0 6px rgba(250, 2, 2, 0.61); }

body::-webkit-scrollbar-thumb {
  background-color: #9b0606;
  outline: 1px solid #6d0404; }
