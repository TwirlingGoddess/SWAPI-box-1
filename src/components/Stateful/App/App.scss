.App {
  text-align: center;
  background-color: black;
  height: 100vh;
  width: 100%; 
  background-image: url('../../../assets/star_wars_kyloren_ren.gif');
  background-repeat: no-repeat;
  background-size: cover
  font-family: 'Nanum Gothic Coding', monospace;
}

body {
  font-family: 'Nanum Gothic Coding', monospace;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg); }
  to {
    transform: rotate(360deg); } }


  body::-webkit-scrollbar {
      width: 1em;
  }
   
  body::-webkit-scrollbar-track {
      -webkit-box-shadow: inset 0 0 6px rgba(250, 2, 2, 0.61);
  }
   
  body::-webkit-scrollbar-thumb {
    background-color: rgb(155, 6, 6);
    outline: 1px solid rgb(109, 4, 4);
  }
  