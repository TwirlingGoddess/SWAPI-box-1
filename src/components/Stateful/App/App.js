import React, { Component } from 'react';
import '../../../reset.css'
import Background from '../Background/Background'
import CardDisplay from '../../Stateless/CardDisplay/CardDisplay'
import Header from '../Header/Header'
import './App.css';

class App extends Component {
  constructor () {
    super()
    this.state = {
      selectedData: [] 
    }
  }

  makeApiCall = async (category) => {
    const url = `https://swapi.co/api/${category}`;
    const fetchURL = await fetch(url)
    const parseObject = await fetchURL.json()
    const cleanData = await this.cleanData(parseObject)
    await this.setState({
      selectedData: cleanData
    })
  }


  cleanData = async (dataObject) => {
    if(dataObject.next.includes('people')){
      return await this.peopleObject(dataObject.results)
    }
    if(dataObject.next.includes('planets')){
      return this.planetObject(dataObject.results)
    }
    if(dataObject.next.includes('vehicles')){
      return await this.vehicleObject(dataObject.results)
    }
  }

  peopleObject = async (parsedData) => {
    const unresolvedPromises = parsedData.map( async(person) => {
      const homeworld = await this.nestedFetch(person.homeworld)
      const homeworldName = homeworld.name
      const homeworldPopulation = homeworld.population
      const specieObj = await this.nestedFetch(person.species)
      const specie = specieObj.name
      const name = person.name;
      return {homeworldName, homeworldPopulation, specie, name}
    })
    return Promise.all(unresolvedPromises)
  }

  planetObject = async (parsedData) => {
    const unresolvedPromises = parsedData.map(async(planet) => {
      const planetName = planet.name;
      const planetTerrain = planet.terrain;
      const planetPopulation = planet.population;
      const planetClimate = planet.climate;
      const residents = await this.residentsFetch(planet.residents)
      return {planetName, planetTerrain, planetPopulation, planetClimate, residents}
    })
    return Promise.all(unresolvedPromises)
  }

  vehicleObject = (parsedData) => {
    const vehicleArray = parsedData.map((vehicle) => {
      const vehicleName = vehicle.name;
      const vehicleModel = vehicle.model;
      const vehicleClass = vehicle.vehicle_class;
      const numberOfPassengers = vehicle.passengers;
      return {vehicleName, vehicleModel, vehicleClass, numberOfPassengers}
    })
    return vehicleArray;
  }

  nestedFetch = async (url) => {
    const fetchURL= await fetch(url)
    const parseObject= await fetchURL.json()
    return parseObject;
  }

  residentsFetch = async (residentsUrls) => {
    const unresolvedPromises = residentsUrls.map(async(url) => {
      const residentURLfetch = await fetch(url)
      const parseResidents = await residentURLfetch.json()
      return parseResidents.name;
    })
    return Promise.all(unresolvedPromises)
  }

  render() {

    if (this.state.selectedData.length){
      return (
        <div>
          <CardDisplay selectedData={this.state.selectedData}/>
        </div>
      )
    }
    
    else if (!this.state.selectedData.length){
      return (
        <div className="App">
          <Header makeApiCall={this.makeApiCall} />   
          <Background />
        </div>
      );
    } 
      
  }
}

export default App;
