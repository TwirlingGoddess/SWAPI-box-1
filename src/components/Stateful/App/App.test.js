import React from 'react';
import App from './App';
import Helper from '../../Helper/Helper'
import { shallow } from 'enzyme';
import { mockPeopleApiResponse, mockPeopleObject } from '../../../_mocks_/mockData';
import 'jest-localstorage-mock'

describe('app', () => {
  let app
  let helper
  it('matches snapshot', () => {
    expect(app).toMatchSnapshot()
  })

  describe('apiCall', () => {

    beforeEach(() => {
      app = shallow(<App />, { disableLifecycleMethods: true })
      helper = new Helper()
      window.fetch = jest.fn()
      helper.sendToLocalStorage = jest.fn()
    })

    it('should update the state', async () => {

      // window.fetch.mockImplementation(() => {
      //   return Promise.resolve({ 
      //     status: 200,
      //     ok: true,
      //     json: () => Promise.resolve( mockPeopleApiResponse )
      //   })
      // })
      // helper.sendToLocalStorage = jest.fn()
      // helper.makeApiCall = jest.fn()

      // await app.instance().apiCall()
    })
  })

  describe('componenDidMount', () => {
    it('should set the state if there is something is local storage', () => {
      app.instance().componentDidMount()
      expect(app.state().favorites).toHaveLength(0)
    })
  })

  describe('FindCard', () => {
    let selectedCard    

    beforeEach(() => {
      
    })
    it('finds the selected card and pushes it to the favorites array', () => {
      selectedCard = {id: "people0", keyList: "people", Homeworld: "Tatooine", Population: "200000", Specie: "Human"} 
      app.instance().findCard(selectedCard)
      expect(app.state().favorites).toHaveLength(1)
    })

    it('calls addToFavorites ', () => {
      app.instance().addCardToFavorites = jest.fn()
      app.instance().findCard(selectedCard)
      expect(app.instance().addCardToFavorites).toHaveBeenCalled()
    })
  })

  describe('addCardFavorites', () => {
    let selectedCard    

    it('checks for duplicates', () => {
      const expectedResult = mockPeopleObject
      app.setState({
        favorites: mockPeopleObject
      })
      selectedCard = {id: "people0", keyList: "people", Homeworld: "Tatooine", Population: "200000", Specie: "Human"} 
      app.instance().addCardToFavorites(selectedCard)
      expect(app.state().favorites).toEqual(expectedResult);
    })
  })

  describe('displayFavorites', () => {
    let selectedCard    

    it('checks for duplicates', () => {
      const expectedResult = mockPeopleObject
      app.setState({
        favorites: mockPeopleObject
      })
      app.instance().displayFavorites()
      expect(app.state().selectedData).toEqual(expectedResult)

    })
  })

})
