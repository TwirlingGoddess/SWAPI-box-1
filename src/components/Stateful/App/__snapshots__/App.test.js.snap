// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app matches snapshot 1`] = `undefined`;
